addons/digest/tests/test_digest.py

```diff
@@ -0,0 +1,75 @@
+from unittest.mock import patch
+import random
+
+from dateutil.relativedelta import relativedelta
+from lxml import html
+
+from odoo.addons.mail.models.mail_mail import MailMail
+
+from odoo import fields
+from odoo.tests import TransactionCase
+
+
+class TestDigestStuff(TransactionCase):
+    def test_digest_numbers(self):
+        self._setup_messages()
+
+        d = self.env['digest.digest'].create({'name': "My Digest", 'kpi_mail_message_total': True})
+
+        # subscribe the demo user so at least one mail gets sent
+        demo = self.env.ref('base.user_demo')
+        d_demo = d.with_env(self.env(user=demo))
+        d_demo.action_subscribe()
+        self.assertTrue(
+            d_demo.is_subscribed,
+            "check the user was subscribed as action_subscribe will silently "
+            "ignore subs of non-employees"
+        )
+
+        # digest creates its mails in auto_delete mode so we need to capture
+        # the formatted body during the sending process
+        body = None
+        def send(self, *args, _original=MailMail.send, **kwargs):
+            nonlocal body
+            body = self.body_html
+            return _original(self, *args, **kwargs)
+
+        with patch.object(MailMail, 'send', autospec=True, wraps=MailMail, side_effect=send):
+            d.action_send()
+        self.assertTrue(body, "a mail has been created for the digest")
+
+        kpi_values = html.fromstring(body).xpath('//*[hasclass("kpi_cell")]//*[hasclass("kpi_value")]')
```

Comment:
Is it a reliable test ? I mean, is it working for example when other KPIs are involved (like having crm and digest) ? Seems a bit low-level and crash-test.

Reply:
It's always going to be low-level since the entire point is to test that the mail is correctly generated (without double-escaping content) and that mail is not exactly conducive to programmatic manipulations. I guess hooking into the kpi_fullname to check that this is the correct "field" could help but only to the extent that the field's label does not change. The alternative is to go and edit the template itself in order to add the field name itself in the output somewhere.